core_types/src/eav.rs

```diff
@@ -287,87 +357,19 @@ impl EntityAttributeValueStorage for ExampleEntityAttributeValueStorage {
 
     fn fetch_eavi(
         &self,
-        entity: Option<Entity>,
-        attribute: Option<Attribute>,
-        value: Option<Value>,
-        index_query: IndexQuery,
+        query: &EaviQuery,
     ) -> Result<BTreeSet<EntityAttributeValueIndex>, HolochainError> {
-        let map = self.storage.read()?;
-        let new_map = map.clone();
-        let filtered: BTreeSet<EntityAttributeValueIndex> = new_map
-            .into_iter()
-            .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.entity(), entity.as_ref()))
-            .filter(|e| {
-                EntityAttributeValueIndex::filter_on_eav_with_prefix(
-                    &e.attribute(),
-                    attribute.as_ref(),
-                    &index_query,
-                )
-            })
-            .filter(|e| EntityAttributeValueIndex::filter_on_eav(&e.value(), value.as_ref()))
-            .filter(|e| {
-                index_query
-                    .start()
-                    .map(|start| start <= e.index())
-                    .unwrap_or_else(|| {
-                        let latest = get_latest(e.clone(), map.clone(), index_query.clone())
-                            .unwrap_or_default();
-                        latest.index() == e.index()
-                    })
-            })
-            .filter(|e| {
-                index_query
-                    .end()
-                    .map(|end| end >= e.index())
-                    .unwrap_or_else(|| {
-                        let latest = get_latest(e.clone(), map.clone(), index_query.clone())
-                            .unwrap_or(EntityAttributeValueIndex::default());
-                        latest.index() == e.index()
-                    })
-            })
-            .collect::<BTreeSet<EntityAttributeValueIndex>>();
-
-        Ok(filtered)
+        let lock = self.storage.read()?;
+        let set = (*lock).clone();
+        let iter = set.iter().cloned();
+        Ok(query.run(iter))
     }
 }
 
-pub fn get_latest(
```

Comment:
The logic of `get_latest` was DRY'd up and moved to `EaviQuery::run` under the `LatestByAttribute` variant of `IndexFilter`